source/common/upstream/original_dst_cluster.cc

```diff
@@ -0,0 +1,145 @@
+  host_set_.addMemberUpdateCb([this](const std::vector<HostSharedPtr>& hosts_added,
+                                     const std::vector<HostSharedPtr>& hosts_removed) -> void {
+    // Update the hosts map
+    for (const HostSharedPtr& host : hosts_removed) {
+      if (host_map_.remove(host)) {
+        ENVOY_LOG(debug, "Removing host {}.", host->address()->asString());
+      }
+    }
+    for (const HostSharedPtr& host : hosts_added) {
+      if (host_map_.insert(host)) {
+        ENVOY_LOG(debug, "Adding host {}.", host->address()->asString());
+      }
+    }
+  });
+}
+
+HostConstSharedPtr
+OriginalDstCluster::LoadBalancer::chooseHost(const LoadBalancerContext* context) {
+  if (context) {
+    const Network::Connection* connection = context->downstreamConnection();
+
+    // The local address of the downstream connection is the original destination address,
+    // if usingOriginalDst() returns 'true'.
+    if (connection && connection->usingOriginalDst()) {
+      const Network::Address::Instance& dst_addr = connection->localAddress();
+
+      // Check if a host with the destination address is already in the host set.
+      HostSharedPtr host = host_map_.find(dst_addr);
+      if (host) {
+        ENVOY_LOG(debug, "Using existing host {}.", host->address()->asString());
+        host->used(true); // Mark as used.
+        return std::move(host);
+      }
+      // Add a new host
+      const Network::Address::Ip* dst_ip = dst_addr.ip();
+      if (dst_ip) {
+        Network::Address::InstanceConstSharedPtr host_ip_port;
+        if (dst_ip->version() == Network::Address::IpVersion::v4) {
+          host_ip_port.reset(
+              new Network::Address::Ipv4Instance(dst_ip->addressAsString(), dst_ip->port()));
+        } else {
+          host_ip_port.reset(
+              new Network::Address::Ipv6Instance(dst_ip->addressAsString(), dst_ip->port()));
+        }
+        // Create a host we can use immediately.
+        host.reset(new HostImpl(info_, info_->name() + dst_addr.asString(), std::move(host_ip_port),
+                                false, 1, ""));
+
+        ENVOY_LOG(debug, "Created host {}.", host->address()->asString());
+        // Add the new host to the map.  We just failed to find it in
+        // our local map above, so insert without checking (2nd arg == false).
+        host_map_.insert(host, false);
+
+        if (ClusterSharedPtr parent = parent_.lock()) {
+          // Dynamic cast below is guaranteed to succeed, as code instantiating the cluster
+          // configuration, that is run prior to this code, checks that an OriginalDstCluster is
+          // always configured with an OriginalDstCluster::LoadBalancer, and that an
+          // OriginalDstCluster::LoadBalancer is never configured with any other type of cluster,
+          // and throws an exception otherwise.
+          Event::Dispatcher& main_dispatcher =
+              dynamic_cast<OriginalDstCluster&>(*parent).dispatcher_;
```

Comment:
I'm not crazy about the use of dynamic_cast here, but I don't see a vastly better way of doing this. Can we do the dynamic_cast in the constructor though? Basically, you can just have a weak_ptr<> to your explicit type and then do the dynamic_cast once. Then you don't even need the ClusterWeakPtr definition at top level.